Add save graph and tensor logging to tensorboard

CatherineSue opened the pull request:
Add a customized tensor scalar to tensorboard by using the
custom_scalar plugin in tensorboard. Each line in the scalar
corresponds to an element in the tensor.

Wrap the tensorboard logging module into a new class `Summary`
in file rllab/misc/tensor_summary.py. It supports both the
simple value and tensor logging. It also saves the
computation graph created by rllab.

To record the tensor into tensorboard, use the
`record_tensor` function in file rllab/misc/logger.py.

Refer to: #39, #38

CatherineSue commented:
The test result of the custom tensor scalar is shown in the following image. The scalar groups tensors by their name_scope. 
![tensorboard](https://user-images.githubusercontent.com/8605658/40505874-7a5fdeb8-5f4a-11e8-9afa-c5112646995a.png)

CatherineSue commented:
Changed member names of class `Summary` in file 
rllab/misc/tensor_summary.py in order to be consistent 
with the class in #58.

ryanjulian commented on rllab/misc/logger.py:
```diff
@@ -3,6 +3,7 @@
 from rllab.misc.tabulate import tabulate
 from rllab.misc.console import mkdir_p, colorize
 from rllab.misc.autoargs import get_all_parameters
+from rllab.misc.tensor_summary import Summary
```
please reorder and group imports to conform to PEP8

https://www.python.org/dev/peps/pep-0008/#imports

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+from rllab.misc.console import mkdir_p
+
+
+class Summary:
```
1. Classes should always have a name matching the file containing them, e.g. `tensor_summary.py` --> `TensorSummary` or `summary.py` --> `Summary`

ryanjulian replied:
I think `TensorBoardOutput` or `TensorBoardOutputFormat` would be a better name for this class.

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+
+class Summary:
+    def __init__(self):
+        self._summary_scale = tf.Summary()
```
what is the purpose of this variable? is it for scalars?

ryanjulian replied:
i.e. should it be called `self._scalar_summary`? If so, you could shorten it to `self._scalars`

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+
+    def record_tensor(self, name, val):
+        scope = str(name).split('/', 1)[0]
+        if scope not in self._scope_tensor.keys():
```
i think in python 3 you can just say `if scope not in self._scope_tensor:`

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+        else:
+            if name not in self._scope_tensor[scope]:
+                self._scope_tensor[scope].append(name)
+            val = np.array(val).reshape(-1)
```
shouldn't this be outside the `else` to update the value of `scope/name` in all cases?

ryanjulian replied:
val.flatten()?

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+                self._scope_tensor[scope].append(name)
+            val = np.array(val).reshape(-1)
+            for i in range(len(val)):
+                self.record_scale(name + '/' + str(i), val[i])
```
optional: preserve the dimensions and make this "(i, j, k, ...)" by iterating over the dimensions

ryanjulian replied:
record_scalar()?

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+            if name not in self._scope_tensor[scope]:
+                self._scope_tensor[scope].append(name)
+            val = np.array(val).reshape(-1)
+            for i in range(len(val)):
```
```python
for i, v in enumerate(val):
    # v == val[i]
```

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+        self._writer.add_graph(tf.get_default_graph())
+        self._writer.flush()
+
+    def record_scale(self, key, val):
```
`record_scalar()`?

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+            for i in range(len(val)):
+                self.record_scale(name + '/' + str(i), val[i])
+
+    def dump_tensor(self, step):
```
hmm how does this differ from calling `dump_scale()`?

ryanjulian replied:
what happens if i call this without calling `self.layout_tensor()`?

CatherineSue replied:
`dump_tensor()` is the same as `dump_scale()`, which is now refactored as `dump_scalor()`. And if you call this without calling `self.layout_tensor()`, there will be no customized scalar about the tensor. Should I refector `self.layout_tensor()` to `self.dump_tensor()`?

ryanjulian replied:
I think there should be one API call which, when called by a user, guarantees that the latest tensors show up on TensorBoard. To conform to the rest of the API, it should probably be called `dump_tensors()`

ryanjulian replied:
Also, spelling, it is spelled `scalar` `dump_scalar()` `scalars`, etc.

ryanjulian commented on rllab/misc/tensor_summary.py:
```diff
@@ -0,0 +1,89 @@
+            self._writer = tf.summary.FileWriter(dir_name)
+            self._default_step = 0
+            assert self._writer is not None
+            print("tensorboard data will be logged into:", dir_name)
```
Please use the text logger for this rather than print.

ryanjulian reviewed (changes requested):
This looks pretty good.

I think it needs some clarifications to the variable names and to the overall API. I'm not sure which API calls are public and which are private, and when I can call which API.

Side note: please remove don't include files in a PR which are unrelated to the PR. It looks like your changes got mixed with old changes from a different PR.

To avoid this, start with a fresh branch off of `integration` for each new PR. 

It looks like you might not be tracking the latest code in integration.

You can setup a different remote to make sure your github branch is up-to-date with the central one
```sh
git remote add ryanjulian git@github.com:ryanjulian/rllab.git
git fetch ryanjulian
```

To make sure your github matches the central github, use this recipe.
```sh
git checkout integration
git fetch ryanjulian
git reset --hard integration ryanjulian/integration
git push -f origin integration
```

Then when you start working on a new feature, do this.
```sh
git checkout integration
git checkout -b my_great_new_feature_branch
```

More info: https://gist.github.com/markreid/12e7c2203916b93d23c27a263f6091a0

Commit:
Support for tensor logging to tensorboard

Add a customized tensor scalar to tensorboard by using the
custom_scalar plugin in tensorboard. Each line in the scalar
corresponds to an element in the tensor.

Wrap the tensorboard logging module into a new class `Summary`
in file rllab/misc/tensor_summary.py. It supports both the
simple value and tensor logging. It also saves the
computation graph created by rllab.

To record the tensor into tensorboard, use the
`record_tensor` function in file rllab/misc/logger.py.

Refer to: #39, #38

CatherineSue commented:
Rebased the repo and refactored the code in tensorboard_output.py. Sorry about the earlier messy commits.

Changes I made:

1. Regroup the imports
2. Change the class name `Summary` to `TensorBoardOutput`.
3. Refactor the name of the class member. Set the private member function names to `_*`, majorly the dump function.
4. Remove unrequired class member.
5. Preserve the tensor dimensions and iterate over dimensions. Make the tag name to i,j,k,... * 
*Can't include brackets in the tag of scalars because of tensorboard cursom_scalar doesn't support them

ryanjulian commented on rllab/misc/logger.py:
```diff
@@ -1,21 +1,22 @@
 from enum import Enum
```
nit: alphabetical order for all imports

ryanjulian commented on rllab/misc/tensorboard_output.py:
```diff
@@ -0,0 +1,88 @@
+
+import numpy as np
+import tensorflow as tf
+from tensorboard import summary as summary_lib
```
nit: alphabetical order for all imports

ryanjulian reviewed (approved):
Fix the nits and submit!

CatherineSue commented:
Needs improvement:
1. There will be scalars in both tab Scalar and Customized Scalar in tensorboard. Tried to find a way to show tensors only in Customized Scalar but failed. 
2. The custom_scalar plugin don't group the same layout together. In a result, if we have a log_dir which has several runs, the result of the customized scalar is like this:
<img width="1205" alt="screen shot 2018-05-25 at 4 54 59 pm" src="https://user-images.githubusercontent.com/8605658/40570128-7c96da72-603c-11e8-9bcd-ac45c8fdbba6.png">
There are two trig/bar:0 scalars.

ryanjulian commented:
Hmm okay. Investigate and let me know what you find out. Done is better than perfect :)

CatherineSue commented:
The custom_scalar plugin firstly writes the layout proto into a custom_scalar_config inside the run file. As a result, each run has a custom_scalar_config in its run file. When we start tensorboard at the top-level dir, it recursively finds runs within the tensorboard_dir. Thus, there will be duplicate charts, like the one above.

It seems that the original designer doesn't support the idea of each config per run (https://github.com/tensorflow/tensorboard/pull/664):
> (I think per-run config files should be ignored ... what would a per-run config file mean?)

To deal with this issue, I pertain a global custom_scalar_config in the top-level dir. When there is a new run started, I merge the new run's layout proto into the global layout proto, so that there will be no duplicate fields anymore. When tensorboard starts, it only gets one custom_scalar_config.

e.g. The file inside the `custom_scalar_config` contains the global custom_scalar_config.
<img width="481" alt="screen shot 2018-05-26 at 1 15 06 am" src="https://user-images.githubusercontent.com/8605658/40574152-7b213e4e-6082-11e8-9755-bee06e6c7a60.png">

And the tensorboard result is like this.
<img width="1440" alt="screen shot 2018-05-26 at 1 15 31 am" src="https://user-images.githubusercontent.com/8605658/40574166-a8c59462-6082-11e8-9e88-d77de934cb1f.png">

ryanjulian commented:
Great job!
